packages/server/src/modules/app/tests/request.ts

```diff
@@ -0,0 +1,70 @@
+import supertest, { SuperTest, Test } from 'supertest';
+
+import application from '@application';
+import { TestRequest } from '@types';
+import { User } from '@chess-tent/models';
+
+const { generateApiToken } = application.service;
+
+class Request implements TestRequest {
+  private request: SuperTest<Test>;
+  private method: 'get' | 'post';
+  private url: string;
+  private cookies: string[];
+  private data?: Object | null;
+
+  constructor() {
+    this.request = supertest(application.service.router);
+    this.url = process.env.API_BASE_PATH || '';
+    this.method = 'get';
+    this.cookies = [];
+  }
+
+  init() {
+    return new Request();
+  }
+
+  get(url: string) {
+    this.method = 'get';
+    this.url += url;
+    return this;
+  }
+
+  post(url: string) {
+    this.method = 'post';
+    this.url += url;
+    return this;
+  }
+
+  setAuthorization(user: User) {
+    this.cookies.push(`token=${generateApiToken(user)}`);
+
+    return this;
+  }
+
+  setCookies(cookies: string[]) {
+    this.cookies = [...this.cookies, ...cookies];
+    return this;
+  }
+
+  setBody(newData: Object) {
+    this.data = newData;
+    return this;
+  }
```

Comment:
Can you please make test request immutable by making all of the methods return a new object? :pray:  Your future self and me will be thanking you.
Example:
```
    return {...this, data: newData } as TestRequest;
``` 
tather than
```
    this.data = newData;
    return this;
```

Reply:
@aleksmarkovic there is a lib used in the project called `immerJS`. It's used to create the immutable behaviour. I briefly looked at the docs, it seems it supports classes.

https://immerjs.github.io/immer/complex-objects/

This should be the right way to do it.

Reply:
@brama051 Why would we need a new object on each method if one object is used per request?

Reply:
Think of the usage of it: Let's say you initialize a testRequest in one place. There's baseRequest with auth token and what not. Current implementation makes each call on the method change the object. So you write a set of tests and it's all well because you remembered what's the stat of the object so you continue updating and maybe even overwriting obj attributes (not being aware it's a bad practice). 

Some time later, new requirement comes in and you want to cover it with a test. It belongs to the existing test suite you already worked on, so you add a test somewhere in the middle of your test suite. Then you realize that the rest of the tests are failing now. You look into the functionality, why would it even have an effect on rest of the tests, then you debug, spend your valuable time only to find out that you updated HTTP method to POST so now all the subsequent tests need to be updated.

**Immutability is love. Default to it and use mutability only when there's no other choice.**

Reply:
That is totally logical and I understand.
So even if I am before each request creating a new request object, somebody could still somewhere forget to define
`beforeEach(() => {
  request = application.test.request.init();
});
`

and then we would have the same problem.

I'll try Braco's library, won't that affect performance?

Reply:
I don't think we would have the same problem. 
And what is the purpose of `init()`? 
Why should application.test.request contain the object? 

If this is your test, and TestRequest is immutable, you can initialize your base request without beforeEach().
```
const baseRequest: TestRequest = new Request().get().setAuthorization();


describe('testiram' ()=> {
  it('da se kasnije ne nerviram' () => {
    baseRequest
      .setUrl('recepti')
      .post()
      .setBody({ query: 'krumpir salata' })
  });
  
  it('da bez brige pivu diram' () => {
    
    // ovaj ce request biti get, a ne post ako je baseRequest immutable.
    baseRequest.setUrl('recepti/?filter=krumpir%20salata') 
      .run()
  });
   
}); 

```

Reply:
@brama051 @MrBr check this out: https://github.com/MrBr/chess-tent/pull/15/commits/93f6ea9b57abe8eb6f468c6816790230c623fd6b

Each method is `producing` a new copy on its call and the base object is not touched. Do we need to worry about those creations?

Reply:
🚚 <-- garbage collector

Reply:
@aleksmarkovic the part about the immer looks good.

Reply:
That's it! 👍

Reply:
You can approve then